library/ssl_srv.c

```diff
@@ -776,6 +776,127 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
 }
 #endif /* MBEDTLS_SSL_ALPN */
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP)
+static int ssl_parse_use_srtp_ext( mbedtls_ssl_context *ssl,
+                                   const unsigned char *buf,
+                                   size_t len )
+{
+    mbedtls_ssl_srtp_profile client_protection = MBEDTLS_TLS_SRTP_UNSET;
+    size_t i,j;
+    size_t profile_length;
+    uint16_t mki_length;
+    /*! 2 bytes for profile length and 1 byte for mki len */
+    const size_t size_of_lengths = 3;
+
+    /* If use_srtp is not configured, just ignore the extension */
+    if( ssl->conf->dtls_srtp_profile_list == NULL ||
+        ssl->conf->dtls_srtp_profile_list_len == 0 )
+    {
+        return( 0 );
+    }
+
+    /* RFC5764 section 4.1.1
+     * uint8 SRTPProtectionProfile[2];
+     *
+     * struct {
+     *   SRTPProtectionProfiles SRTPProtectionProfiles;
+     *   opaque srtp_mki<0..255>;
+     * } UseSRTPData;
+
+     * SRTPProtectionProfile SRTPProtectionProfiles<2..2^16-1>;
+     */
+
+    /*
+     * Min length is 5: at least one protection profile(2 bytes)
+     *                  and length(2 bytes) + srtp_mki length(1 byte)
+     * Check here that we have at least 2 bytes of protection profiles length
+     */
+    if( len < 2 )
+    {
+        mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL,
+                                        MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE );
+        return( MBEDTLS_ERR_SSL_BAD_HS_CLIENT_HELLO );
+    }
+
+   ssl->dtls_srtp_info.chosen_dtls_srtp_profile = MBEDTLS_TLS_SRTP_UNSET;
+
+    /* first 2 bytes are protection profile length(in bytes) */
+    profile_length = ( buf[0] << 8 ) | buf[1];
+    buf += 2;
+
+    /* check the buffer size: at least profiles + profile and mki length */
+    if( profile_length + size_of_lengths > len ||
```

Comment:
profile_length is of type size_t but we use it to load a 16bits value read in the input buffer. So unless we built on a 16 bits processor having thus size_t mapped on a 16 bits unsigned integer, there is no possible overflow.

Is mbedtls being built on 16 bits processors? Unless this https://tls.mbed.org/discussions/crypto-and-ssl/help-with-ecc-and-bignum is no more relevant, I assume no.

I'll change it anyway just in case at some point 16 bits processors are supported.

Reply:
It's ok to assume that `int` and `size_t` are 32-bit types or larger. There's already plenty of code in the library that assumes that adding a few 16-bit or 24-bit values doesn't overflow.

([The non-support for 16-bit platforms could stand to be better documented](https://github.com/ARMmbed/mbedtls/issues/3323).)

Reply:
Yes, we can assume that `int` and `size_t` are 32 bit or larger. I missed that we load a 16 bit value in `profile_length`.

Reply:
@jeannotlapin Thank you for changing it! This could save us a couple of false alarms in the future if someone makes the same mistake I did.